pkg/rtc/transport.go

```diff
@@ -325,30 +335,43 @@ func (t *PCTransport) createAndSendOffer(options *webrtc.OfferOptions) error {
 	// if restart is requested, and we are not ready, then continue afterwards
 	if iceRestart {
 		if t.pc.ICEGatheringState() == webrtc.ICEGatheringStateGathering {
-			t.logger.Debugw("restart ICE after gathering")
+			t.params.Logger.Debugw("restart ICE after gathering")
 			t.restartAfterGathering = true
 			return nil
 		}
-		t.logger.Debugw("restarting ICE")
+		t.params.Logger.Debugw("restarting ICE")
 	}
 
-	// when there's an ongoing negotiation, let it finish and not disrupt its state
-	if t.negotiationState == negotiationStateClient {
+	if iceRestart && t.negotiationState != negotiationStateNone {
 		currentSD := t.pc.CurrentRemoteDescription()
-		if iceRestart && currentSD != nil {
-			t.logger.Debugw("recovering from client negotiation state")
+		if currentSD == nil {
+			// never received an answer from client, create a new peer connection
+			t.params.Logger.Infow("no previous answer from client on ICE restart, creating a new PC")
+			t.pc.Close()
+			if err := t.createPeerConnection(); err != nil {
+				t.params.Logger.Errorw("could not create new PC on ICE restart", err)
+				return err
+			} else {
+				options.ICERestart = false
+			}
+		} else {
+			// recover by re-applying the last answer
+			t.params.Logger.Infow("recovering from client negotiation state on ICE restart")
 			if err := t.pc.SetRemoteDescription(*currentSD); err != nil {
 				prometheus.ServiceOperationCounter.WithLabelValues("offer", "error", "remote_description").Add(1)
 				return err
 			}
-		} else {
-			t.logger.Debugw("skipping negotiation, trying again later")
+		}
+	} else {
```

Comment:
if `iceRestart == true && t.negotiationState == negotiationStateNone`, we would've missed setting options.ICERestart = true

Reply:
`iceRestart` is `true` only when `options.ICERestart` is `true`. So, it will fall through and have that option set.

Reply:
ah you are right. it's already passed in. sorry got confused.